### DISCUSSION TITLE
How to get helpful feedback on bad code exceptions

### DISCUSSION THREAD

**Discussion by Beakerboy** [Beakerboy/VBA-Linter#21] [General]
@Xenios91  Do you know if there is a way to easily reproduce
`antlr4-parse vba.g4 startRule _file_` using the API? The moment I call `InputStream(bad_code)` I just get an exception. I’d love to print out the same helpful error message the the antlr4-parse tool gives if someone tries to lint some code with a small error in it, like accidentally typing “End Sub” instead of “End Function”.

**Comment by Xenios91** [comment]
So you want to be able to detect invalid code right? Have you looked at the error node within the tree? It is visited when something goes wrong, such as forgetting to end a loop or forgetting to end a function or something as you mentioned. If you want smart errors going to that and traversing it's children or parents within the tree may give additional context to know what went wrong.

**Comment by Beakerboy** [comment]
Thanks. I tried to start with just calling vbaParser.getNumberOfSyntaxErrors(), but that seems to just be returning 0 even when there is an obvious error like:

```vba
Public Function()
End Sub
```

do you know if I need to add an error listener for any of this?

**Reply by Beakerboy** [reply]
Ah! I have to actually call `parser.startRule()` before it actually parses anything. I got it to tell me there are errors, now to mess around with getting the line and column of the errors and look at the messages.

**Reply by Xenios91** [reply]
Glad you got it figured out, sorry ive been really busy so im slow to respond.
